[REST API] Reset tracks and preferences after application passwords logout

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #8235
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR ensures that analytics and prefs resetting are done both for wpcom-based and REST API-based account logout.

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hafizrahman committed:
Add: Dispatch sign out action during application passwords logout, too.

hafizrahman committed:
Add: handle SIGN_OUT as proper causeOfChange from SIGN OUT action.

hafizrahman committed:
Remove extraneous brackets.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/AppInitializer.kt:
```diff
@@ -304,8 +304,8 @@ class AppInitializer @Inject constructor() : ApplicationLifecycleListener {
     fun onAccountChanged(event: OnAccountChanged) {
-        val isLoggedOut = event.causeOfChange == null && event.error == null
-        if (!accountStore.hasAccessToken() && isLoggedOut) {
+        val isLoggedOut = event.causeOfChange == AccountAction.SIGN_OUT && event.error == null
```
As far as I can see, any log out action from the app eventually generates `AccountAction.SIGN_OUT` from FluxC as the `causeOfChange`, never `null`.

From the initial commit that added it (https://github.com/woocommerce/woocommerce-android/commit/68595f425956e9147aafbbda5766d6209194ac8d), I couldn't find the reason why `event.causeofChange == null` is used to check logged out state.

Because of that, in the current `trunk` state, the "Reset analytics" and "Wipe user-specific preferences" part of the code below never gets called. This change is meant to fix that.

hafizrahman replied:
This is still a draft PR but I could use a confidence check on this @hichamboushaba @JorgeMucientes , thank you!

hichamboushaba replied:
I'm on my phone, I'll just give a few hints, and tomorrow we can discuss this with more details 🙂 

> Because of that, in the current trunk state, the "Reset analytics" and "Wipe user-specific preferences" part of the code below never gets called. This change is meant to fix that.

Nice catch 👍, I believe I'm the one who updated FluxC to return the correct cause there, but I see this broke the logic here 🤦.

But regarding the main change here, I feel like the `AccountActionBuilder.newSignOutAction()` is meant for a WPCom account logout, since it simply clears the WPCom token, and using it just to trigger the `AccountAction` doesn't seem like the right thing to do.
When I wrote the issue for this bug, I thought about moving this cleanup logic to our `AccountRepository`, IMO it'll make more sense to be placed there in the `logout` function, since every logout call is made through the `AccountRepository`, what do you think about this?

hafizrahman replied:
Thanks for checking, I see what you meant. My approach was to make sure any type of logout gets handled here, but your suggestion to put the cleanup in the repository instead makes sense. I actually had the cleanup separately added in the repository before but thought it's tidier to reuse the code here.

I'll work on it.

hafizrahman committed:
Optimize code.

hafizrahman committed:
Refactor: put cleanup action inside AccountRepository instead.

hafizrahman commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/AccountRepository.kt:
```diff
@@ -48,16 +53,18 @@ class AccountRepository @Inject constructor(
             if (event.isError) {
                 WooLog.e(
-                    SITE_PICKER,
+                    LOGIN,
```
Just a smaller change for what I think is the more proper tag.

hafizrahman committed:
Refactor: return value nesting.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/login/AccountRepository.kt:
```diff
@@ -45,19 +50,22 @@ class AccountRepository @Inject constructor(
                 )
                 false
             } else {
                 dispatcher.dispatch(SiteActionBuilder.newRemoveWpcomAndJetpackSitesAction())
```
We delete sites when signing out from WordPress.com account, but we don't do it for site credentials, I think we should. WDYT about moving this line to the `cleanup` function, and then deleting all sites, not just the Jetpack ones? (just replacing `newRemoveWpcomAndJetpackSitesAction` with `newRemoveAllSitesAction`).

hafizrahman committed:
Refactor: delete all sites for all logout cases.

Also opt to use `newRemoveAllSitesAction` to delete everything instead of `newRemoveWpcomAndJetpackSitesAction`

hichamboushaba reviewed (approved):
Nice work @hafizrahman :shipit:
